lib/trino-parquet/src/main/java/io/trino/parquet/reader/ChunkedInputStream.java

```diff
@@ -0,0 +1,151 @@
+ * This abstraction is used because the page size is unknown until the page header is read
+ * and page header and page data can be split between two or more byte chunks.
+ */
+public final class ChunkedInputStream
+        extends InputStream
+{
+    private final Iterator<? extends ChunkReader> chunks;
+    private ChunkReader currentChunkReader;
+    private BasicSliceInput current;
+
+    public ChunkedInputStream(Collection<? extends ChunkReader> chunks)
+    {
+        requireNonNull(chunks, "chunks is null");
+        checkArgument(!chunks.isEmpty(), "At least one chunk is expected but got none");
+        this.chunks = chunks.iterator();
+        readNextChunk();
+    }
+
+    public Slice getSlice(int length)
+            throws IOException
+    {
+        if (length == 0) {
+            return Slices.EMPTY_SLICE;
+        }
+        ensureOpen();
+        while (!current.isReadable()) {
+            checkArgument(chunks.hasNext(), "Requested %s bytes but 0 was available", length);
+            readNextChunk();
+        }
+        if (current.available() >= length) {
+            return current.readSlice(length);
+        }
+        // requested length crosses the slice boundary
+        byte[] bytes = new byte[length];
+        try {
+            readFully(this, bytes, 0, bytes.length);
+        }
+        catch (IOException e) {
+            throw new RuntimeException("Failed to read " + length + " bytes", e);
+        }
+        return Slices.wrappedBuffer(bytes);
+    }
+
+    @Override
+    public int read(byte[] b, int off, int len)
+            throws IOException
+    {
+        checkPositionIndexes(off, off + len, b.length);
+        if (len == 0) {
+            return 0;
+        }
+        ensureOpen();
+        while (!current.isReadable()) {
+            if (!chunks.hasNext()) {
+                return -1;
+            }
+            readNextChunk();
+        }
+
+        return current.read(b, off, len);
+    }
+
+    @Override
+    public int read()
+            throws IOException
+    {
+        ensureOpen();
+        while (!current.isReadable() && chunks.hasNext()) {
+            readNextChunk();
+        }
+
+        return current.read();
+    }
+
+    @Override
+    public int available()
+            throws IOException
+    {
+        ensureOpen();
+        return current.available();
+    }
+
+    @Override
+    public void close()
+    {
+        if (current == null) {
+            // already closed
+            return;
+        }
+        currentChunkReader.free();
+
+        while (chunks.hasNext()) {
+            chunks.next().free();
+        }
+        current = null;
+    }
+
+    private void ensureOpen()
+            throws IOException
+    {
+        if (current == null) {
+            throw new IOException("Stream closed");
+        }
+    }
+
+    private void readNextChunk()
+    {
+        if (currentChunkReader != null) {
+            currentChunkReader.free();
```

Comment:
Maybe we should also free `current`?

Reply:
WDYM free current? set it to `null`? it s being replaced by another value below.

Reply:
Yes, set it to `null`. It will not be replaced if exception will be thrown.